xtra-libp2p-rollover/src/v_2_0_0/taker.rs

```diff
@@ -84,13 +85,42 @@ impl<E, O> Actor<E, O> {
                 v_2_0_0::PROTOCOL,
             ))
             .await
-            .context("Endpoint is disconnected")?
-            .context("No connection to peer")?
-            .await
-            .context("Failed to open substream")?;
+            .context("Endpoint is disconnected")?;
+
+        let substream = match substream.context("No connection to peer") {
+            Ok(substream) => substream,
+            Err(e) => {
+                self.emergency_drop_connection(peer_id, &e).await?;
+                return Err(e);
+            }
+        };
```

Comment:
🔧 This error means that you don't have a connection to the peer. You don't need to drop the connection!

Reply:
But we previously saw instances where we did not recover from this. I think it can happen that the endpoint does not notice that the connection is broken and then it will never notify the dialer actor to re-connect and we keep running into this problem.

Reply:
There is only one place where we `remove` from `controls: HashMap<PeerId, (yamux::Control, Tasks)>`, which is when we call `drop_connection` which also notifies all subscribers that the connection was dropped. It's here:

https://github.com/itchysats/itchysats/blob/fb75a75349454685fe4c8fd932e86bfe8d72ff76/xtra-libp2p/src/endpoint.rs#L257-L260

Here you're matching on a `NoConnection` error, which can only be emitted if the `PeerId` doesn't exist in the `controls` map. And what you are then doing is sending a `Disconnect` message which will _always do nothing_ because the `PeerId` doesn't exist in the `controls` map according to the error you just got from the `Endpoint`.